haystack_experimental/components/extractors/llm_metadata_extractor.py

```diff
@@ -0,0 +1,172 @@
+    We have also had significant co-brand momentum in CEMEA. First, we launched a new co-brand card in partnership with Qatar Airways, British Airways and the National Bank of Kuwait. Second, we expanded our strong global Marriott relationship to launch Qatar's first hospitality co-branded card with Qatar Islamic Bank. Across the United Arab Emirates, we now have exclusive agreements with all the leading airlines marked by a recent agreement with Emirates Skywards.
+    And we also signed an inaugural Airline co-brand agreement in Morocco with Royal Air Maroc. Now newer digital issuers are equally
+    ------------------------
+    output:
+    {"entities": [{"entity": "Visa", "entity_type": "company"}, {"entity": "Alaska Airlines", "entity_type": "company"}, {"entity": "Qatar Airways", "entity_type": "company"}, {"entity": "British Airways", "entity_type": "company"}, {"entity": "National Bank of Kuwait", "entity_type": "company"}, {"entity": "Marriott", "entity_type": "company"}, {"entity": "Qatar Islamic Bank", "entity_type": "company"}, {"entity": "Emirates Skywards", "entity_type": "company"}, {"entity": "Royal Air Maroc", "entity_type": "company"}]}
+    #############################
+    -Real Data-
+    ######################
+    entity_types: [company, organization, person, country, product, service]
+    text: {{input_text}}
+    ######################
+    output:
+    '''
+
+    docs = [
+        Document(content="deepset was founded in 2018 in Berlin, and is known for its Haystack framework"),
+        Document(content="Hugging Face is a company founded in Paris, France and is known for its Transformers library")
+    ]
+
+    extractor = LLMMetadataExtractor(prompt=NER_PROMPT, expected_keys=["entities"])
+    extractor.run(documents=docs)
+    >> {'documents': [
+        Document(id=.., content: 'deepset was founded in 2018 in Berlin, and is known for its Haystack framework',
+        meta: {'entities': [{'entity': 'deepset', 'entity_type': 'company'}, {'entity': 'Berlin', 'entity_type': 'city'},
+              {'entity': 'Haystack', 'entity_type': 'product'}]}),
+        Document(id=.., content: 'Hugging Face is a company founded in Paris, France and is known for its Transformers library',
+        meta: {'entities': [
+                {'entity': 'Hugging Face', 'entity_type': 'company'}, {'entity': 'Paris', 'entity_type': 'city'},
+                {'entity': 'France', 'entity_type': 'country'}, {'entity': 'Transformers', 'entity_type': 'product'}
+                ]})
+           ]
+       }
+    >>
+    ```
+    """ # noqa: E501
+    def __init__(
+            self,
+            prompt: str,
+            expected_keys: List[str],
+            model: Optional[str] = None,
+            raise_on_failure: bool = True
+    ):
+        self.prompt = prompt
+        self.builder = PromptBuilder(prompt)
+        self.generator = OpenAIGenerator() if model is None else OpenAIGenerator(model=model)
```

Comment:
Instead of just choosing the OpenAIGenerator, could we make the generator an init param? We have situations where need to use other generators for performance reasons or deployment reasons. E.g. We need to use AzureOpenAIGenerator or the AWSBedrockGenerator for enterprise requirements. 

Also the way this is currently specified it's not possible to pass any additional params to the Generator component like the name of the env variable that stores the API key or generation params like setting temp to 0.

Reply:
It's not so easy to generalise this to any Generator, since there's no abstraction. I will for now, just add these 3:

- OpenAIGenerator
- AzureOpenAIGenerator
- AmazonBedrockGenerator

any other you would like to have?

Reply:
Just double checking, but generalizing here would just be about having an expanded type for Generator right? I believe all Generators follow the same run method signature or are we worried that isn't respected?

Reply:
Otherwise, I think those three + the GeminiGenerator should be good for now.

Reply:
> Just double checking, but generalizing here would just be about having an expanded type for Generator right? I believe all Generators follow the same run method signature or are we worried that isn't respected?

Yes, but that means making this PR dependent on a change on Haystack core - but we can definitely think about it.

> Otherwise, I think those three + the GeminiGenerator should be good for now.
Alright, adding that one as well

Reply:
```
    def __init__(
        self,
        prompt: str,
        input_text: str,
        expected_keys: List[str],
        raise_on_failure: bool = True,
        generator: Union[OpenAIGenerator, AzureOpenAIGenerator, AmazonBedrockGenerator],
        model: Optional[str] = None,
        generation_kwargs: Optional[Dict[str, Any]] = None,
    ):
```

If we have an `__init__()` like this wouldn't make sense that the generator is already initalised and passed to the `LLMMetadataExtractor()`'s `__init()__` and we remove the` model` and `generation_kwargs`.

Alternatively we change the generator to be an `Enum` and depending on it's value we initialised on of the possible generators it inside the `LLMMetadataExtractor`

Do you see any other possibilty?

The whole issue here is wanting to be free to use any possible `Generator` but not having an abstraction to represent it.

Reply:
> If we have an __init__() like this wouldn't make sense that the generator is already initalised and passed to the LLMMetadataExtractor()'s __init()__ and we remove the model and generation_kwargs.

Yeah I think this makes the most sense. I'd prefer this approach.

Reply:
Also to quickly pop in, rolling back on this really limits the usability of the component for some clients. I recognize the nesting of components is a broader topic, so I wanted to ask if we don't do this now (in this PR), when we might expect a decision to be made on that topic?

Reply:
I know that nesting components is a big topic, but we could use this PR to first experiment with nested components in the context of @sjrl specific use case. Given that this is the experimental package, it's the best place to try it.

Reply:
Another option than sticking to hard types would be introducing a [protocol](https://typing.readthedocs.io/en/latest/spec/protocol.html). That way you don't need to rely on any types of the main package as long as you can specify the interface.

Reply:
@tstadel that's a good point and we discussed about the need of Abstractions for Generators, it would make cases like this one easier to deal with. It would be beneficial to have something similar like we have for DocumentStores. 

The extra work involves guaranteeing that every LLM underneath works with the defined interface and that you can easily switch LLMs without anything breaking, but I think is the right direction.